### DISCUSSION TITLE
Keyword _HC Date/Time Comparisons

### DISCUSSION THREAD

**Discussion by jbk68** [cortexrd/Knack-Toolkit-Library#113] [General]
Hey Norm, started new discussion as you requested...  Please let me know when you finish the Date/Time Comparisons.  I think this might solve my issues with using the Knack's unreliable scheduled task system that I've had nothing but problems with for months.

My goal is to hide action links both on the grid and detail views.  I'd like to be able to choose the time to hide/show (ie 17:00 hide, 08:00 show) different links.

Thanks again for for all you do for the Knack community... JON

**Comment by CSWinnall** [comment]
Hi Jon

While there is not a keyword at the moment.  You could use Make (formerly integromat) to schedule a scenario to run at the time you want to turn the link/column on and off to update a field to a 1 or a zero then use ktlCond to actually hide the column.  This is a workaround until Norm comes up with an alternative


Craig

**Reply by jbk68** [reply]
Hey Craig, I'll look into Make as a "temporary" alternative.  They can't be any worse than Knack's schedule task system.

I hope Norm comes up with a solution soon.  Thanks for the suggestion... JON

**Comment by CSWinnall** [comment]
We use Make for all sorts of things with Knack.  I think it's worth looking
into.

Craig

**Reply by jbk68** [reply]
Hey Craig, are you experiencing login issues with Knack apps?  I've been getting texts from users stating that can't log into our apps (desktop and mobile) made using Knack.  It appears that something with KTL is causing issue.  If I disable KTL and try logging into app using a device that was having issues, works just fine.  Enable KTL and it screen freezes just prior to login screen.

**Comment by CSWinnall** [comment]
Hi Jon

Yes we are we have raised it with Norm

Craig

**Reply by jbk68** [reply]
I reported issue to Knack prior to me kinda figuring out it was a Knack issue and here is there response:

We found an error in the browser console referencing a KTL file and the problem appears to be this one https://ctrnd.s3.amazonaws.com/Lib/KTL/KTL_Start.js

Based on the error message, it appears that the file doesn't have defined the function of ktl.sysInfo.setCfg. Temporarily commenting your JS code should fix this problem, while Norm finds what happened to their code.

**Reply by jbk68** [reply]
I've disabled / commented out KTL code and you don't realize what you got til its gone.  LOL!!!

I've done a lot of customizing of our apps, as I'm sure you have, with KTL features.

JON

**Comment by cortexrd** [comment]
Guys,

**The _hc now works with ktlCond and time ranges !!!**

This is the syntax:

`_hc=[col1, col2, col3], [ktlCond, in, 17:00-08:00, ktlNow]`

This will hide the 3 columns - both fields and action links will work - if the local time is between 5pm and 8am.

If you have international users, you may prefer to use the `ktlNowUTC` option.  Then all comparisons will be made using the UTC time.

Enjoy,
Norm

**Reply by jbk68** [reply]
Sweet...  Will give it a try.  I looking to replace the built in scheduling task system for Knack.

Thanks Norm... JON

**Reply by jbk68** [reply]
Works as expected both for fields and action links in a table.  

Is there a way to make this work for fields and action links on a details view page? 

Once again my goal is to get away from the SLOOOOW and unreliable built in scheduling task system in Knack and hide certain action links and show certain fields when these action links are hidden on a details view page.

For example, hide action1 between 18:00 - 7:59 and show a message (field1) in its place states for example "Action Not Available".  Then when 8:00 rolls around, action1 is shown and message (field1) is hidden.  

Is this do able???

**Comment by cortexrd** [comment]
Jon,

I just fixed a bug that prevented that from working with Details views.  It should be good now.
Do a **Ctrl+F5** please.  We're still at 0.17.0.

This is what I have in my details view:

```
_ar=5
_hf=[Email], [ktlCond, in, 17:00-08:08, ktlNow]
_hf=[Total Sales], [ktlCond, in, 08:09-17:00, ktlNow]
```

It hides the Email action or the Total Sales field, in complementary time ranges.  So it's always one or the other that is visible.

Does that work for you also?
Norm

**Reply by jbk68** [reply]
**WORKS GREAT!!!**

 Using KTL and your awesome improvements, I can start transitioning from **NOT** using scheduled tasks built into Knack for hiding/showing certain action links and fields based on time.  You have no idea of the issues this will resolve.

Thank you so much... JON

**Reply by cortexrd** [reply]
Thanks Jon!  Very happy to read that.

Spread the news - tell others in the Knack Forums to go for the KTL !!!

Cheers,
Norm

**Reply by jbk68** [reply]
So Norm, I know that using the below, I can hide the (3) fields listed between 6pm - 6:00am:

_hf=[Field_1, Field_2, Field_3 ], [ktlCond, in, 18:00-6:00, ktlNow]

But I'd like to show the same (3) fields for a certain period of time based on a customer zone (ie Eastern)
One of the data fields that I use for my customers is a time zone field which I've been using to update data using Knack's scheduling task and allows certain fields/action links (ie email) to be shown during certain hours based on their time zone.

Example: 
** show fields 1-3 from 7am to 5pm if field_100 = "Eastern"    ** show action1 from 7am to 5pm if field_100 "Eastern"
** show fields 1-3 from 8am to 6pm if field_100 = "Central"     ** show action1 from 8am to 6pm if field_100 "Central"
** show fields 1-3 from 9am to 7pm if field_100 = "Mountain"   ** show action1 from 9am to 7pm if field_100 "Mountain"

Is this possible for both a field and action link?  And is it possible to show a field rather that hide it ?

I know that you just put in the times to hide, but it's a little easier to put in when to show rather than when to hide. 

Thank you sir... JON

**Reply by cortexrd** [reply]
1. For the time zones, it's not possible as is.  I am running out of time to develop this any further at the moment.
But... you can use the ktlRoles to achieve your goal.

```
_hf=[Field_1, Field_2, Field_3 ], [ktlCond, in, 17:01-6:59, ktlNow], [ktlRoles, Easter]
_hf=[Field_1, Field_2, Field_3 ], [ktlCond, in, 18:01-7:59, ktlNow], [ktlRoles, Central]
_hf=[Field_1, Field_2, Field_3 ], [ktlCond, in, 19:01-8:59, ktlNow], [ktlRoles, Mountain]
```

And assign the proper role to each account.

2. No, there's no way to show something, only to hide them at the moment.  Maybe later.  Implementing this would require significant additional work, and I have no time.  We need more collaborators!  Just to maintain the documentation up to date is quite something.

ND

**Reply by jbk68** [reply]
Okay, thank you for the info Norm... JON

**Comment by cortexrd** [comment]
Hi to all,

The ktlCond has been updated.  It can now compare time spans using the ktlNow (or ktlNowUTC) with the **out** operator.

For example:

`_hf=[Email], [ktlCond, out, 17:00-08:00, ktlNow]`

Will hide the email field if local time is not between 8am and 5pm.

It has been released in **KTL v0.19.7**

Enjoy,
Norm

**Reply by jbk68** [reply]
Hey Norm, will give this a try and thank you for KTL.  It's made Knack so much better... JON
